app/models/fire_cloud_client.rb

```diff
@@ -1332,6 +1336,61 @@ def delete_user_from_billing_project(project_id, role, email)
     end
   end
 
+  ##
+  # PET SERVICE ACCOUNT METHODS
+  # these methods reference the SAM API directly for issuing pet service account tokens/json keyfiles
+  # NOTE: the FireCloudClient instance calling these methods must be initialized using a User object
+  # for authentication purposes, otherwise downstream calls will return 403
+  ##
+
+  # create a new instance of FireCloudClient, but use a pet service account keyfile
+  # will only work with users that have been registered in Terra
+  #
+  # * *params*
+  #   - +user+: (User) => User object from which access tokens are generated
+  #   - +project+: (String) => Default GCP Project to use (can be overridden by other parameters)
+  #
+  # * *return*
+  #   - +FireCloudClient+ instance, or nil if user has not registered with Terra
+  def self.new_with_pet_sa(user, project)
+    # create a temporary client in order to retrieve the user's pet service account keyfile
+    tmp_client = self.new(user, project)
+    if tmp_client.registered?
+      pet_service_account_json = tmp_client.get_pet_service_account_key(project)
+      self.new(user, project, pet_service_account_json)
+    else
+      nil
+    end
+  end
+
+  # issue an access_token for a user's pet service account in the requested project
+  #
+  # * *params*
+  #   - +project_name+ (String) => Name of a FireCloud billing project in which pet service account resides
+  #
+  # * *returns*
+  #   - +String+ pet service account OAuth2 access_token
+  def get_pet_service_account_token(project_name)
+    path = "https://sam.dsde-prod.broadinstitute.org/api/google/v1/user/petServiceAccount/#{project_name}/token"
+    # normal scopes, plus RO access for storage objects (removes unnecessary billing scope from GOOGLE_SCOPES)
+    token = process_firecloud_request(:post, path, GOOGLE_SCOPES.to_json)
+    token.gsub(/\"/, '') # gotcha for removing escaped quotes in response body
+  end
+
+  # get JSON keyfile contents for a user's pet service account in the requested project
+  # response from this API call can be passed to FireCloudClient.new(user, project_name, service_account_contents)
+  # to create an instance of FireCloudClient that is able to call GCS methods as the user in the request project
+  #
+  # * *params*
+  #   - +project_name+ (String) => Name of a FireCloud billing project in which pet service account resides
+  #
+  # * *returns*
+  #   - +Hash+ parsed contents of pet service account JSON keyfile
+  def get_pet_service_account_key(project_name)
+    path = "https://sam.dsde-prod.broadinstitute.org/api/google/v1/user/petServiceAccount/#{project_name}/key"
```

Comment:
To avoid needing coordinated updates, it's worth abstracting "https://sam.dsde-prod.broadinstitute.org/api/google/v1/user/petServiceAccount", e.g. to an instance variable, and using that here and in `get_pet_service_account_token`.

```suggestion
    path = "#{@pet_sa_url}/#{project_name}/key"
```

Reply:
f3422a0